File: 065-滑动窗口的最大值/065-maxInWindows.py
# -*- coding:utf-8 -*-
class Solution:
    def maxInWindows(self, num, size):
        # write code here
        if size==0:
            return []
        deque=[]
        ans=[]
        for idx,val in enumerate(num):
            if idx>=size and idx-size==deque[0]:
                deque.pop(0)
            while deque and num[deque[-1]]<=val:
                deque.pop()
            deque.append(idx)
            if idx>=size-1:
                ans.append(num[deque[0]])
        return ans

File: 065-滑动窗口的最大值/065-maxInWindows.cpp
class Solution {
public:
    vector<int> maxInWindows(const vector<int>& num, unsigned int size)
    {
        vector<int> ans;
        deque<int> dq;
        if(size==0){
            return ans;
        }
        for(int i =0;i<num.size();i++){
            if(i>=size && i-size==dq.front()){
                dq.pop_front();
            }
            while(!dq.empty() && num[dq.back()]<=num[i]){
                dq.pop_back();
            }
            dq.push_back(i);
            if (i>=size-1){
                ans.push_back(num[dq.front()]);
            }
        }
        return ans;
    }
};
